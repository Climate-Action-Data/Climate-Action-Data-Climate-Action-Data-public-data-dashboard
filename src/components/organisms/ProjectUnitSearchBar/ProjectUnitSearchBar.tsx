import { CloseIcon } from '@/components/atoms/CloseIcon/CloseIcon'
import { Button, Container, Flex, Input, InputGroup, InputRightElement } from '@chakra-ui/react'
import { useTranslation } from 'react-i18next'
import React, { ChangeEvent, FC, useState } from 'react'
import { SearchIcon } from '@/components/atoms/SearchIcon/SearchIcon'
import { useRouter } from 'next/navigation'
import { ESearchParams } from '@/@types/ProjectSearchResult'
import { useActions } from '@/overmind'

interface ProjectUnitSearchBarProp {
  isProjectSearch?: boolean
}

const ProjectUnitSearchBar: FC<ProjectUnitSearchBarProp> = (props) => {
  const { isProjectSearch } = props
  const { setKeywordSearch } = useActions().searchFilters
  const [searchInput, setSearchInput] = useState(``)
  const router = useRouter()
  const { t } = useTranslation(`search`)

  const handleOnClear = () => {
    setSearchInput(``)
  }

  const handleInputOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchInput(event.target.value)
  }

  const handleOnSearch = () => {
    const searchParams = new URLSearchParams()
    searchParams.append(ESearchParams.KEYWORD, searchInput)

    setKeywordSearch(searchInput)

    if (isProjectSearch) {
      router.push(`/search/projects`)
    } else {
      router.push(`/search/units`)
    }
  }

  const handleOnKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === `Enter`) {
      handleOnSearch()
    }
  }

  return (
    <Container variant={`searchBar`} width={[`100%`, null, null, `888px`]}>
      <Flex>
        <InputGroup variant={`searchInput`}>
          {searchInput.length !== 0 && (
            <InputRightElement>
              <CloseIcon onClick={handleOnClear} data-testid={`search-bar-clear`} />
            </InputRightElement>
          )}
          <Input
            placeholder={isProjectSearch ? t(`searchProjectsByKeywordsPlaceholder`) : t(`searchUnitsByKeywordsPlaceholder`)}
            value={searchInput}
            onChange={handleInputOnChange}
            onKeyDown={handleOnKeyDown}
          />
        </InputGroup>
        <Button rightIcon={<SearchIcon width={`16px`} height={`16px`} />} variant={`accentPrimary32`} data-testid={`search-bar-search`} onClick={handleOnSearch} marginLeft={`8px`}>
          {t(`search`)}
        </Button>
      </Flex>
    </Container>
  )
}

export default ProjectUnitSearchBar
