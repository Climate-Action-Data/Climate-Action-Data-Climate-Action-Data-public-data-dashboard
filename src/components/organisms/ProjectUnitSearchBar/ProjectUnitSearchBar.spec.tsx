import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ProjectUnitSearchBar from '@/components/organisms/ProjectUnitSearchBar/ProjectUnitSearchBar'

const mockPush = jest.fn()
jest.mock(`next/navigation`, () => ({
  ...jest.requireActual(`next/navigation`),
  useRouter: () => ({ push: mockPush }),
  useSearchParams: () => ({ get: () => `testParams` }),
}))

it(`renders correctly for project search`, () => {
  const { container } = render(<ProjectUnitSearchBar isProjectSearch={true} />)
  expect(container).toMatchSnapshot()
})

it(`renders correctly for unit search`, () => {
  const { container } = render(<ProjectUnitSearchBar isProjectSearch={false} />)
  expect(container).toMatchSnapshot()
})

it(`renders correctly for project search with no params, add a search value and submits a search`, async () => {
  const { container } = render(<ProjectUnitSearchBar isProjectSearch={true} />)
  await userEvent.type(screen.getByRole(`textbox`), `hello`)
  await userEvent.click(screen.getByTestId(`search-bar-clear`))
  await userEvent.click(screen.getByTestId(`search-bar-search`))
  expect(mockPush).toHaveBeenCalledWith(`/search/projects?keyword=`)
  expect(container).toMatchSnapshot()
})

it(`renders correctly and perform a search on enter`, async () => {
  const { container } = render(<ProjectUnitSearchBar isProjectSearch={false} />)
  await userEvent.type(screen.getByRole(`textbox`), `hello`)
  await userEvent.keyboard(`{enter}`)
  expect(mockPush).toHaveBeenCalledWith(`/search/units?keyword=hello`)
  expect(container).toMatchSnapshot()
})
