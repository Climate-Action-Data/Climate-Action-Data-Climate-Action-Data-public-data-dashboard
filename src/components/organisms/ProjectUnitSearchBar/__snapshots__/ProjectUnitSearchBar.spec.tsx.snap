// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly and perform a search on enter 1`] = `
<div>
  <div
    class="chakra-container css-10oehrs"
  >
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <div
          class="chakra-input__right-element css-dp9r5f"
          variant="searchInput"
        >
          <svg
            class="chakra-icon css-n059si"
            data-testid="search-bar-clear"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.0005 13.0543L6.92737 18.1274C6.78892 18.2658 6.61489 18.3367 6.40527 18.3399C6.19567 18.3431 6.01844 18.2723 5.87357 18.1274C5.72869 17.9825 5.65625 17.8069 5.65625 17.6005C5.65625 17.3941 5.72869 17.2184 5.87357 17.0736L10.9466 12.0005L5.87357 6.92738C5.73511 6.78893 5.66427 6.61489 5.66107 6.40527C5.65786 6.19568 5.72869 6.01844 5.87357 5.87358C6.01844 5.72869 6.19407 5.65625 6.40047 5.65625C6.60687 5.65625 6.78251 5.72869 6.92737 5.87358L12.0005 10.9467L17.0736 5.87358C17.212 5.73511 17.3861 5.66428 17.5957 5.66108C17.8053 5.65786 17.9825 5.72869 18.1274 5.87358C18.2723 6.01844 18.3447 6.19408 18.3447 6.40048C18.3447 6.60688 18.2723 6.78251 18.1274 6.92738L13.0543 12.0005L18.1274 17.0736C18.2658 17.212 18.3367 17.3861 18.3399 17.5957C18.3431 17.8053 18.2723 17.9825 18.1274 18.1274C17.9825 18.2723 17.8069 18.3447 17.6005 18.3447C17.3941 18.3447 17.2184 18.2723 17.0736 18.1274L12.0005 13.0543Z"
              fill="currentColor"
            />
          </svg>
        </div>
        <input
          class="chakra-input css-0"
          placeholder="Search for units by project name/country/standard/unit status"
          value="hello"
        />
      </div>
      <button
        class="chakra-button css-1d6mcmw"
        data-testid="search-bar-search"
        type="button"
      >
        Search
        <span
          class="chakra-button__icon css-1hzyiq5"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1hyqtbe"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.1623 20.2239L13.4084 14.4701C12.9084 14.8829 12.3334 15.206 11.6834 15.4393C11.0334 15.6726 10.361 15.7893 9.66613 15.7893C7.95696 15.7893 6.51045 15.1975 5.32658 14.0139C4.14271 12.8304 3.55078 11.3842 3.55078 9.67552C3.55078 7.96678 4.14256 6.52011 5.32613 5.33549C6.5097 4.15089 7.95585 3.55859 9.66458 3.55859C11.3733 3.55859 12.82 4.15053 14.0046 5.33439C15.1892 6.51826 15.7815 7.96478 15.7815 9.67394C15.7815 10.388 15.6616 11.0701 15.4219 11.7201C15.1821 12.3701 14.8623 12.9355 14.4623 13.4162L20.2411 19.1951C20.3796 19.3309 20.4488 19.4975 20.4488 19.695C20.4488 19.8925 20.3706 20.0695 20.214 20.226C20.0705 20.3695 19.8956 20.4412 19.6892 20.4412C19.4828 20.4412 19.3072 20.3688 19.1623 20.2239ZM9.66613 14.2893C10.9546 14.2893 12.0459 13.8422 12.9402 12.948C13.8344 12.0538 14.2815 10.9624 14.2815 9.67394C14.2815 8.38548 13.8344 7.29413 12.9402 6.39989C12.0459 5.50566 10.9546 5.05854 9.66613 5.05854C8.37766 5.05854 7.28631 5.50566 6.39208 6.39989C5.49786 7.29413 5.05076 8.38548 5.05076 9.67394C5.05076 10.9624 5.49786 12.0538 6.39208 12.948C7.28631 13.8422 8.37766 14.2893 9.66613 14.2893Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders correctly for project search 1`] = `
<div>
  <div
    class="chakra-container css-10oehrs"
  >
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          class="chakra-input css-0"
          placeholder="Search for projects by project name/methodology/country"
          value=""
        />
      </div>
      <button
        class="chakra-button css-1d6mcmw"
        data-testid="search-bar-search"
        type="button"
      >
        Search
        <span
          class="chakra-button__icon css-1hzyiq5"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1hyqtbe"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.1623 20.2239L13.4084 14.4701C12.9084 14.8829 12.3334 15.206 11.6834 15.4393C11.0334 15.6726 10.361 15.7893 9.66613 15.7893C7.95696 15.7893 6.51045 15.1975 5.32658 14.0139C4.14271 12.8304 3.55078 11.3842 3.55078 9.67552C3.55078 7.96678 4.14256 6.52011 5.32613 5.33549C6.5097 4.15089 7.95585 3.55859 9.66458 3.55859C11.3733 3.55859 12.82 4.15053 14.0046 5.33439C15.1892 6.51826 15.7815 7.96478 15.7815 9.67394C15.7815 10.388 15.6616 11.0701 15.4219 11.7201C15.1821 12.3701 14.8623 12.9355 14.4623 13.4162L20.2411 19.1951C20.3796 19.3309 20.4488 19.4975 20.4488 19.695C20.4488 19.8925 20.3706 20.0695 20.214 20.226C20.0705 20.3695 19.8956 20.4412 19.6892 20.4412C19.4828 20.4412 19.3072 20.3688 19.1623 20.2239ZM9.66613 14.2893C10.9546 14.2893 12.0459 13.8422 12.9402 12.948C13.8344 12.0538 14.2815 10.9624 14.2815 9.67394C14.2815 8.38548 13.8344 7.29413 12.9402 6.39989C12.0459 5.50566 10.9546 5.05854 9.66613 5.05854C8.37766 5.05854 7.28631 5.50566 6.39208 6.39989C5.49786 7.29413 5.05076 8.38548 5.05076 9.67394C5.05076 10.9624 5.49786 12.0538 6.39208 12.948C7.28631 13.8422 8.37766 14.2893 9.66613 14.2893Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders correctly for project search with no params, add a search value and submits a search 1`] = `
<div>
  <div
    class="chakra-container css-10oehrs"
  >
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          class="chakra-input css-0"
          placeholder="Search for projects by project name/methodology/country"
          value=""
        />
      </div>
      <button
        class="chakra-button css-1d6mcmw"
        data-testid="search-bar-search"
        type="button"
      >
        Search
        <span
          class="chakra-button__icon css-1hzyiq5"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1hyqtbe"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.1623 20.2239L13.4084 14.4701C12.9084 14.8829 12.3334 15.206 11.6834 15.4393C11.0334 15.6726 10.361 15.7893 9.66613 15.7893C7.95696 15.7893 6.51045 15.1975 5.32658 14.0139C4.14271 12.8304 3.55078 11.3842 3.55078 9.67552C3.55078 7.96678 4.14256 6.52011 5.32613 5.33549C6.5097 4.15089 7.95585 3.55859 9.66458 3.55859C11.3733 3.55859 12.82 4.15053 14.0046 5.33439C15.1892 6.51826 15.7815 7.96478 15.7815 9.67394C15.7815 10.388 15.6616 11.0701 15.4219 11.7201C15.1821 12.3701 14.8623 12.9355 14.4623 13.4162L20.2411 19.1951C20.3796 19.3309 20.4488 19.4975 20.4488 19.695C20.4488 19.8925 20.3706 20.0695 20.214 20.226C20.0705 20.3695 19.8956 20.4412 19.6892 20.4412C19.4828 20.4412 19.3072 20.3688 19.1623 20.2239ZM9.66613 14.2893C10.9546 14.2893 12.0459 13.8422 12.9402 12.948C13.8344 12.0538 14.2815 10.9624 14.2815 9.67394C14.2815 8.38548 13.8344 7.29413 12.9402 6.39989C12.0459 5.50566 10.9546 5.05854 9.66613 5.05854C8.37766 5.05854 7.28631 5.50566 6.39208 6.39989C5.49786 7.29413 5.05076 8.38548 5.05076 9.67394C5.05076 10.9624 5.49786 12.0538 6.39208 12.948C7.28631 13.8422 8.37766 14.2893 9.66613 14.2893Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders correctly for unit search 1`] = `
<div>
  <div
    class="chakra-container css-10oehrs"
  >
    <div
      class="css-k008qs"
    >
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          class="chakra-input css-0"
          placeholder="Search for units by project name/country/standard/unit status"
          value=""
        />
      </div>
      <button
        class="chakra-button css-1d6mcmw"
        data-testid="search-bar-search"
        type="button"
      >
        Search
        <span
          class="chakra-button__icon css-1hzyiq5"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-1hyqtbe"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.1623 20.2239L13.4084 14.4701C12.9084 14.8829 12.3334 15.206 11.6834 15.4393C11.0334 15.6726 10.361 15.7893 9.66613 15.7893C7.95696 15.7893 6.51045 15.1975 5.32658 14.0139C4.14271 12.8304 3.55078 11.3842 3.55078 9.67552C3.55078 7.96678 4.14256 6.52011 5.32613 5.33549C6.5097 4.15089 7.95585 3.55859 9.66458 3.55859C11.3733 3.55859 12.82 4.15053 14.0046 5.33439C15.1892 6.51826 15.7815 7.96478 15.7815 9.67394C15.7815 10.388 15.6616 11.0701 15.4219 11.7201C15.1821 12.3701 14.8623 12.9355 14.4623 13.4162L20.2411 19.1951C20.3796 19.3309 20.4488 19.4975 20.4488 19.695C20.4488 19.8925 20.3706 20.0695 20.214 20.226C20.0705 20.3695 19.8956 20.4412 19.6892 20.4412C19.4828 20.4412 19.3072 20.3688 19.1623 20.2239ZM9.66613 14.2893C10.9546 14.2893 12.0459 13.8422 12.9402 12.948C13.8344 12.0538 14.2815 10.9624 14.2815 9.67394C14.2815 8.38548 13.8344 7.29413 12.9402 6.39989C12.0459 5.50566 10.9546 5.05854 9.66613 5.05854C8.37766 5.05854 7.28631 5.50566 6.39208 6.39989C5.49786 7.29413 5.05076 8.38548 5.05076 9.67394C5.05076 10.9624 5.49786 12.0538 6.39208 12.948C7.28631 13.8422 8.37766 14.2893 9.66613 14.2893Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
